File: 수학/[6588번]골드바흐의 추측/[6588번]골드바흐의 추측.py
# https://www.acmicpc.net/problem/6588

import sys

prime = [i for i in range(1000001)]
prime[1] = 0

for i in range(2, 1001):
	if prime[i] != 0:
		for j in range(i * 2, 1000001, i):
			prime[j] = 0

while True:
	n = int(sys.stdin.readline().rstrip())
	if n == 0:
		break
	for i in range(3, n + 1):
		if prime[i] == 0 or prime[n - i] == 0:
			continue
		print(str(n) + " = " + str(i) + " + " + str(n - i))
		break
		if i == n:
			print("Goldbach's conjecture is wrong.")

File: 수학/[6588번]골드바흐의 추측/[6588번] 골드바흐의 추측.java
import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.OutputStreamWriter;
import java.util.Arrays;

public class Main {
	public static void main(String[] args) throws IOException {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
		boolean[] prime = new boolean[1000001];
		prime[0] = true;
		prime[1] = true;
		for (int i = 2; i < Math.sqrt(1000001); i++) {
			if (prime[i]) continue;
			int j = 2;
			while (i * j < 1000001) {
				prime[i * j++] = true;
			}
		}		
		while (true) {
			int n = Integer.parseInt(br.readLine());
			if (n == 0) break;
			int p = -1;
			for (int i = 2; i < (n / 2) + 1; i++) {
				if (prime[i]) continue;
				if (!prime[n - i]) {
					p = i;
					break;
				}
			}
			if (p == -1) bw.write("Goldbach's conjecture is wrong.\n");
			else bw.write(String.valueOf(n) + " = " + String.valueOf(p) +" + "+ String.valueOf(n - p) + "\n");
			
		}
		bw.flush();
		bw.close();
	}

}
